File: trunk/arduino/libraries/ArduRCT_EventManager/ArduRCT_Analog.cpp
/*
 * ArduRCT_Analog - Handles analog channel changes
 *
 * Copyright (c) 2010 Laurent Wibaux <lm.wibaux@gmail.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */


#include "ArduRCT_Analog.hpp"
#include "ArduRCT_EventManager.h"

ArduRCT_Analog::ArduRCT_Analog(uint8_t pin, uint8_t averaging, uint16_t resolution) {
    _pin = pin;
    _value = 0;
    _averaging = averaging;
    if (_averaging == 0) averaging = 1;
    _resolution = resolution;
    _previousValue = ANALOG_NO_VALUE;
    pinMode(pin, INPUT);
}

uint8_t ArduRCT_Analog::getPin() {
    return _pin;
}

ArduRCT_Analog *ArduRCT_Analog::getNext() {
    return _next;
}

void ArduRCT_Analog::setNext(ArduRCT_Analog *next) {
    _next = next;
}

uint16_t ArduRCT_Analog::updateValue() {
    _value = 0;
    for (uint8_t i=0; i<_averaging; i++) _value += analogRead(_pin);
    _value = _value/_averaging;
    if (_resolution != ANALOG_HARDWARE_RESOLUTION) {
        uint32_t value32 = _value;
        value32 = value32 * _resolution / ANALOG_HARDWARE_RESOLUTION;
        _value = value32;
    }
    return _value;
}

int16_t ArduRCT_Analog::getChange() {
    int16_t change;
    if (_previousValue == ANALOG_NO_VALUE) change = 0;
    else change = _value - _previousValue;
    _previousValue = _value;
    return change;
}

uint16_t ArduRCT_Analog::getValue() {
    return _value;
}

File: trunk/java/ArduRCT_Simulator/src/com/google/code/ardurct/libraries/eventManager/ArduRCT_Analog.java
package com.google.code.ardurct.libraries.eventManager;

import com.google.code.ardurct.hardware.Analog;
import com.google.code.ardurct.libraries.IArduinoDefines;

public class ArduRCT_Analog implements IEventDefines, IArduinoDefines {

	
	ArduRCT_Analog _next = null;
	int _pin;
	int _resolution; 
	int _previousValue;
	int _value;
	int _history[] = new int[ANALOG_MAX_AVERAGING];
	int _historyIndex;
	int _averaging;
	
	public ArduRCT_Analog(int pin) {
		this(pin, ANALOG_AVERAGING, ANALOG_RESOLUTION_12B);
	}


	public ArduRCT_Analog(int pin, int averaging) {
		this(pin, averaging, ANALOG_RESOLUTION_12B);
	}

	public ArduRCT_Analog(int pin, int averaging, int resolution) {
		_pin = pin;
		_value = 0;
		_historyIndex = -1;
		_previousValue = ANALOG_NO_VALUE;
		_averaging = averaging;
		if (_averaging > ANALOG_MAX_AVERAGING) _averaging = ANALOG_MAX_AVERAGING;
		else if (_averaging < 1) _averaging = 1; 
		_resolution = resolution;
		pinMode(pin, INPUT);
	}

	public int getPin() {
		return _pin;
	}
	
	public ArduRCT_Analog getNext() {
		return _next;
	}
	
	public void setNext(ArduRCT_Analog next) {
		_next = next;
	}

	public int updateValue() {
		if (_historyIndex == -1) {
			_history[0] = analogRead(_pin);
			for (int i=1; i<_averaging; i++) _history[i] = _history[0];
		} else _history[_historyIndex] = analogRead(_pin);
		_historyIndex ++;
		if (_historyIndex == _averaging) _historyIndex = 0;
		_value = 0;
		for (int i=0; i<_averaging; i++) _value += _history[i];
		_value = _value/_averaging;
		if (_resolution != ANALOG_HARDWARE_RESOLUTION) {
			int value32 = _value;
			value32 = value32 * _resolution / ANALOG_HARDWARE_RESOLUTION;
			_value = value32;
		}
		return _value;
	}
	
	public int getChange() {
		int change;
		if (_previousValue == ANALOG_NO_VALUE) change = 0;
		else change = _value - _previousValue;
		_previousValue = _value;
		return change;
	}

	public int getValue() {
		return _value;
	}

	private int analogRead(int pin) {
		return Analog.read(pin);
	}
	
	public void pinMode(int pin, int value) {
	}

}
